python/compress-image/test_main.py

```diff
@@ -0,0 +1,359 @@
+
+    @unittest.skipIf(not (secret.API_KEY_KRAKENIO and secret.SECRET_API_KEY_KRAKENIO), "No KrakenIO API Key or Secret Key")
+    def test_krakenio_time_out(self):
+        with patch("main.requests.post") as mock_post:
+            mock_post.side_effect = requests.exceptions.ReadTimeout
+            with self.assertRaises(requests.exceptions.ReadTimeout):
+                main.krakenio_impl({
+                    "api_key": secret.API_KEY_KRAKENIO,
+                    "api_secret_key": secret.SECRET_API_KEY_KRAKENIO,
+                    "decoded_image": IMAGE_1KB
+                })
+            mock_post.assert_called_once()
+
+
+class TestValidateRequest(unittest.TestCase):
+    @parameterized.expand([
+        [
+            {
+                "payload": {
+                    "provider": "tinypng",
+                    "image": str(base64.b64encode(pathlib.Path(
+                        secret.IMAGE_1KB).read_bytes()), "utf-8")
+                },
+                "variables": {"API_KEY": secret.API_KEY_TINYPNG},
+            },
+            {
+                "provider": "tinypng",
+                "api_key": secret.API_KEY_TINYPNG,
+                "decoded_image": IMAGE_1KB
+            }
+        ],
+        [
+            {
+                "payload": {
+                    "provider": "krakenio",
+                    "image": str(base64.b64encode(pathlib.Path(
+                        secret.IMAGE_1KB).read_bytes()), "utf-8")
+                },
+                "variables": {
+                    "API_KEY": secret.API_KEY_KRAKENIO,
+                    "SECRET_API_KEY": secret.SECRET_API_KEY_KRAKENIO
+                }
+            },
+            {
+                "provider": "krakenio",
+                "api_key": secret.API_KEY_KRAKENIO,
+                "api_secret_key": secret.SECRET_API_KEY_KRAKENIO,
+                "decoded_image": IMAGE_1KB
+            }
+        ]
+    ])
+
+    @unittest.skipIf(not (secret.API_KEY_KRAKENIO and secret.SECRET_API_KEY_KRAKENIO and secret.API_KEY_TINYPNG), "No Tinypng Key and KrakenIO Key")
+    def test_validate_request(self, got, expected):
+        req = MyRequest({
+            "payload": got["payload"],
+            "variables": got["variables"]
+            })
+        self.assertEqual(main.validate_request(req), expected)
+
+    @parameterized.expand([
+        [
+            {
+                "payload": {},
+                "variables": {},
+            },
+            ("Missing payload")
+        ],
+        [
+            {
+                "payload": {"provider": "IMNOTAPROVIDER", "image": ""},
+                "variables": {"API_KEY": "1234567"},
+            },
+            ("Invalid provider.")
+        ],
+        [
+            {
+                "payload": {"provider": "krakenio", "image": ""},
+                "variables": {},
+            },
+            ("Missing variables.")
+        ],
+        [
+            {
+                # Missing api key tiny
```

Comment:
The comment and the error string match. The comment doesn't add much that isn't already in the error.